sdks/go/pkg/beam/core/metrics/metrics.go

```diff
@@ -0,0 +1,447 @@
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package metrics handles Beam metrics.
+//
+// Metrics in the Beam model are uniquely identified by a namespace, a name,
+// and the PTransform context in which they are used. Further, they are
+// reported as a delta against the bundle being processed, so that overcounting
+// doesn't occur if a bundle needs to be retried.
+//
+// Metric containers aren't initialized until their first mutation, which
+// follows from the Beam model design, where metrics are only sent for a bundle
+// if they have changed. This is particularly convenient for distributions which
+// means their min and max fields can be set to the first value on creation
+// rather than have some marker of uninitialized state, which would otherwise
+// need to be checked for on every update.
+//
+// Metric values are implemented as lightweight proxies of the user provided
+// namespace and name. This allows them to be declared globally, and used in
+// any ParDo. To handle reporting deltas on the metrics by bundle, metrics
+// are keyed by bundleID,PTransformID,namespace, and name, so metrics that
+// are identical except for bundles are treated as distinct, effectively
+// providing per bundle deltas, since a new value container is used per bundle.
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/util/bundle"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/util/ptransform"
+	"github.com/apache/beam/sdks/go/pkg/beam/log"
+	"github.com/apache/beam/sdks/go/pkg/beam/model/fnexecution_v1"
+	"github.com/golang/protobuf/ptypes"
+)
+
+// userMetricker knows how to convert it's value to a Metrics_User proto.
+type userMetricker interface {
+	toProto() *fnexecution_v1.Metrics_User
+}
+
+type metricName struct {
+	namespace, name string
+}
+
+func (n metricName) String() string {
+	return fmt.Sprintf("%s.%s", n.namespace, n.name)
+}
+
+func validateName(mn metricName) {
+	if len(mn.name) == 0 || len(mn.namespace) == 0 {
+		panic(fmt.Sprintf("namespace and name are required to be non-empty, got %q and %q", mn.namespace, mn.name))
+	}
+}
+
+type metricKey struct {
+	name               metricName
+	bundle, ptransform string
+}
+
+var (
+	metricMu sync.RWMutex
+	// metricStorage is a map of BundleIds to PtransformIds to metrics.
+	metricStorage = make(map[string]map[string]map[metricName]userMetricker)
+	counters      = sync.Map{}
+	distributions = sync.Map{}
+	gauges        = sync.Map{}
+)
+
+type metricType uint8
```

Comment:
Suggestion: if we don't reuse distribution for counters, then it seems we don't need a metricType at all. There is a lot of manual type tracking in this code, which makes it harder to understand.

Reply:
Done.